fplll/hlll.cpp

```diff
@@ -0,0 +1,636 @@
+        // Precompute dR[0]: R[0]^2 * delta = dR[0] * 2^(2*row_expo[0])
+        compute_dR(0);
+        // Precompute eR[0]: R[0] * eta = eR[0] * 2^row_expo[0]
+        compute_eR(0);
+
+        // Set R[1] to b[1] (bf[1] and other usefull variables were swaped previously)
+        m.refresh_R(1);
+        k = 1;
+      }
+      else
+      {
+        // Size reduce b[k - 1]
+        k--;
+        // Since b[k] was not changed, a previous computation of R[k][0..k-1] can be used instead of
+        // recomputing R[k][0..k-1], which is the only interesting part to begin the size reduction
+        // of b[k]
+        m.recover_R(k);
+      }
+    }
+  }
+}
+
+template <class ZT, class FT> bool HLLLReduction<ZT, FT>::lovasz_test(int k)
+{
+#ifndef MODIFIED_LOVASZ_TEST
+  // This Lovasz test is the one proposed in [MSV, ISSAC'09]
+  //
+  // Prior, we used another test, which is dR[k-1].cmp(R(k, k - 1)^2 + R(k, k)^2) <= 0, if R(k, k)
+  // is known (which is not the case here at this step of the computation, but can be computed
+  // thanks to sqrt(sum_{i=k}^{i<n}R(k, i)^2) (indices must be checked)). In the prior version,
+  // since R(k, k)^2 was known, we directly used the test. However, the test was probably not as
+  // accurate as what we hope. However, since this formula is used in hplll
+  // (https://github.com/gilvillard/hplll/releases), this can maybe be retested. An example of
+  // matrices that was not HLLL reduced because of a fail in this test can be generated thanks to
+  // latticegen -randseed 122 r 300 30000.
+  // Such a test can be activate by compiling with -DMODIFIED_LOVASZ_TEST
+  //
+  // TODO: this section must be improved and investigated, i.e.:
+  //   * probably other improvement to be done
+  m.get_norm_square_b(ftmp0, k, expo0);  // ||b[k]||^2 = ftmp0 * 2^expo0
+  m.norm_square_R_row(ftmp1, k, 0, k - 1,
+                      expo1);  // sum_{i = 0}^{i < k - 1}R[k][i]^2 = ftmp1 * 2^expo1
+
+  // If this check is false, we need to reenable
+  // ftmp0.mul_2si(ftmp0, expo0 - expo1);
+  FPLLL_DEBUG_CHECK(expo0 == expo1);
+
+  ftmp1.sub(ftmp0, ftmp1);  // ||b[k]||^2 - sum_{i = 0}^{i < k - 1}R[k][i]^2 = ftmp1 * 2^expo1
+
+  expo0 = m.get_row_expo(k - 1);
+
+  // Here, delta * R(k - 1, k - 1)^2 = dR[k-1] * 2^(2*expo0). We want to compare
+  //   delta * R(k - 1, k - 1)^2 <= ||b[k]||^2 - sum_{i = 0}^{i < k - 1}R[k][i]^2
+  //   dR[k-1] * 2^(2*expo0) <= ftmp1 * 2^expo1
+  //   dR[k-1] <= ftmp1 * 2^(expo1 - 2*expo0)
+  ftmp1.mul_2si(ftmp1, expo1 - 2 * expo0);
+#else  // MODIFIED_LOVASZ_TEST
+  // Modified Lovasz test, following the comment above.
+  // FIXME: probably not maintened.
+
+  m.norm_square_R_row(ftmp1, k, k, m.get_n(),
+                      expo1);       // sum_{i = k}^{i < n}R[k][i]^2 = ftmp1 * 2^expo1
+#ifdef DEBUG
+  m.get_R(ftmp0, k, k - 1, expo0);  // R(k, k - 1) = ftmp0 * 2^expo0
+#else   // DEBUG
+  m.get_R(ftmp0, k, k - 1);  // R(k, k - 1) = ftmp0 * 2^expo0
```

Comment:
There is no expo0 here.

Reply:
I modified the code in bad95bd to explain why this depend on DEBUG. Maybe it is not a good idea, since returning or not expo0 is not related to an efficiency problem.